### PR TITLE
Simplify PHPStan extension

### PR DESCRIPTION
**Pull request by ondrejmirtes**
> 221 additions and 925 deletions. 

Nice, right? :) The main motivation for me to contribute this is for Carbon to stop using non-public facing PHPStan classes like BuiltinMethodReflection (which will be removed in PHPStan 2.0).

Thanks to ClosureTypeFactory, which allows us to create useful PHPStan Reflection objects out of Closure objects, this greatly simplified the PHPStan extension related to macros.

ClosureTypeFactory has been contributed to PHPStan by Can Vural, to support similar macros analysis in Larastan (Laravel), more than 2 years ago.

I hope you like this.

### CODE REVIEW

**Comment by ondrejmirtes** [comment]
I can bring back tests/PHPStan/MacroTest.php in some form if you'd want.

**Comment by ondrejmirtes** [comment]
I'm now waiting for your feedback before finalizing this PR.

File: src/Carbon/PHPStan/MacroMethodReflection.php [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    public function getThrowType(): ?Type
+    {
+        return null;
+    }
+
+    public function hasSideEffects(): TrinaryLogic
+    {
+        return TrinaryLogic::createMaybe();
+    }
+
```
Comment (kylekatarnls) [inline src/Carbon/PHPStan/MacroMethodReflection.php:122]:
```suggestion
```

File: src/Carbon/PHPStan/MacroMethodReflection.php [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    private $methodName;
+
+    /** @var ParametersAcceptor */
+    private $macroClosureType;
+
+    /** @var bool */
+    private $static;
+
+    /** @var string|null */
+    private $docComment;
```
Comment (kylekatarnls) [inline src/Carbon/PHPStan/MacroMethodReflection.php:37]: If you prefer, we dropped PHP < 8.1 on master branch:
```suggestion
    private ClassReflection $declaringClass;
    private string $methodName;
    private ParametersAcceptor $macroClosureType;
    private bool $static;
    private ?string $docComment;
```

File: src/Carbon/PHPStan/MacroExtension.php [hunk 3]
```diff
@@ -29,62 +35,94 @@
...
-            $builtinMacro->isFinal(),
-            $supportAssertions ? null : $builtinMacro->getDocComment(),
-            $supportAssertions ? Assertions::createEmpty() : null,
-            null,
-            $builtinMacro->getDocComment(),
-            [],
+            $methodName,
+            $closureType,
+            $static,
+            $docComment
```
Comment (kylekatarnls) [inline src/Carbon/PHPStan/MacroExtension.php:125]:
```suggestion
            $docComment,
```

**Review by kylekatarnls — APPROVED** [review]

File: src/Carbon/PHPStan/MacroMethodReflection.php (hunk 2 above)
Comment (ondrejmirtes) [reply src/Carbon/PHPStan/MacroMethodReflection.php:37]: Yeah, sure, I noticed that but on the other hand the old code still used PHPDocs so I thought there was a reason for that.

**Comment by ondrejmirtes** [comment]
And we're green now! 👍 

Please squash the commits on merge, thank you.

**Review by kylekatarnls — APPROVED** [review]

**Comment by kylekatarnls** [comment]
Thank you 🙏 💟

**Pull request merged**